django/contrib/auth/management/__init__.py

```diff
@@ -4,14 +4,104 @@
+
+        # Build a dictionary allowing to do the correspondence between the
+        # current permission codenames and the new <codename, name> values.
+        current_to_new_perms = {}
+        for codename, name in _get_builtin_permissions(current_model._meta):
+            old_codename = re.sub(r'_%s$' % (new_model_lower), '_' + old_model_lower, codename)
+            if not backward:
+                current_to_new_perms[old_codename] = {'codename': codename, 'name': name}
+            else:
+                action = old_codename.split('_' + old_model_lower, 1)[0]
+                current_to_new_perms[codename] = {
+                    'codename': old_codename,
+                    'name': 'Can %s %s' % (action, camel_case_to_spaces(self.old_model)),
+                }
+
+        perms = Permission.objects.db_manager(db).filter(codename__in=current_to_new_perms).order_by('codename')
+        for perm in perms:
+            old_codename = perm.codename
+            perm.codename = current_to_new_perms[old_codename]['codename']
+            perm.name = current_to_new_perms[old_codename]['name']
+            perm.save(update_fields={'codename', 'name'})
+
+    def rename_forwards(self, apps, schema_editor):
+        self._rename(apps, schema_editor)
+
+    def rename_backwards(self, apps, schema_editor):
+        self._rename(apps, schema_editor, backward=True)
+
+
+def inject_rename_permissions_operations(plan=None, apps=global_apps, using=DEFAULT_DB_ALIAS, **kwargs):
+    """
+    Insert a `RenamePermissions` operation after every planned `RenameModel`
+    operation.
+    """
+    if plan is None:
+        return
+
+    # Determine whether or not the Permission model is available.
+    try:
+        Permission = apps.get_model('auth', 'Permission')
+    except LookupError:
+        available = False
+    else:
+        if not router.allow_migrate_model(using, Permission):
+            return
+        available = True
+
+    for migration, backward in plan:
+        if ((migration.app_label, migration.name) == ('auth', '0001_initial')):
+            # There's no point in going forward if the initial auth
+            # migration is unapplied as the Permission model will be
+            # unavailable from this point.
+            if backward:
+                break
+            else:
+                available = True
+                continue
+        # The Permission model is not available yet.
+        if not available:
+            continue
+        inserts = []
+        for index, operation in enumerate(migration.operations):
+            if isinstance(operation, migrations.RenameModel):
+                operation = RenamePermissions(
+                    migration.app_label, operation.old_name, operation.new_name
+                )
+                inserts.append((index + 1, operation))
+        for inserted, (index, operation) in enumerate(inserts):
+            migration.operations.insert(inserted + index, operation)
```

Comment:
I'm not so sure about that. If I understand correctly, the plan of a migration should contain a consistent list of operations. These operations are applied in reverse order when migrating backwards. If I alter the index at which the `RenamePermissions` operation is inserted when migrating backwards, this means that this is no longer compliant with the way migrations work (perhaps a signal could be more appropriate in that case if we want to ensure that the "rename" operation is always applied _after_ the `RenameModel` operation).

That said, if I alter the position of the `RenamePermissions` operation as you said, this would allow me to properly support custom verbose names when migrating backwards.

@timgraham, @charettes do you have any opinions on this?